// The Firmament project
// Copyright (c) 2011-2012 Malte Schwarzkopf <malte.schwarzkopf@cl.cam.ac.uk>
//
// BaseMessage is the root element of Firmament's message hierarchy. Any message
// can be treated as a BaseMessage, and this is the type used for transmission
// channels. Subordinate messages are implemented as optional members, and
// need to be registered in the comment below.
//
// **Note:** Protobuf message extensions have no formal inheritance relationship
// to their containers; thus, "subordinate" here is conceptual, rather than in
// terms of C++ inheritance. In particular, a subordinate message CANNOT be
// passed around or otherwise treated as a BaseMessage, including referencing it
// by BaseMessage pointers or references.

package firmament;

// Message registry:
// -------------------
// 001  - TestMessage
// 002  - HeartbeatMessage
// 003  - RegistrationMessage
// 004  - TaskHeartbeatMessage
// 005  - TaskSpawnMessage
// 006  - TaskStateMessage
// 007  - StorageRegistrationMessage
// 008  - TaskInfoRequestMessage
// 009  - TaskInfoResponseMessage
// 010  - TaskDelegationMessage
// 011  - ObtainObjectMessage
// 012  - StorageDiscoverMessage
// 013  - StorageHeartBeatMessage
// 014  - ObjectInfoMessage
// 015  - TaskKillMessage
// 016  - TaskFinalReport   XXX(malte): inconsistent name!

import "messages/test_message.proto";
import "messages/heartbeat_message.proto";
import "messages/registration_message.proto";
import "messages/task_heartbeat_message.proto";
import "messages/task_spawn_message.proto";
import "messages/task_state_message.proto";
import "messages/storage_registration_message.proto";
import "messages/storage_message.proto";
import "messages/task_info_message.proto";
import "messages/task_delegation_message.proto";
import "messages/task_kill_message.proto";
import "base/task_final_report.proto";

message BaseMessage {
  // TODO(malte): Think about having a type identifier in here, and use a
  // special "batch message" type to indicate presence of multiple extensions.
  optional TestMessage test = 1;
  optional HeartbeatMessage heartbeat = 2;
  optional RegistrationMessage registration = 3;
  optional TaskHeartbeatMessage task_heartbeat = 4;
  optional TaskSpawnMessage task_spawn = 5;
  optional TaskStateMessage task_state = 6;
  optional StorageRegistrationMessage storage_registration = 7;
  optional TaskInfoRequestMessage task_info_request = 8;
  optional TaskInfoResponseMessage task_info_response = 9;
  optional TaskDelegationRequestMessage task_delegation_request = 10;
  optional TaskDelegationResponseMessage task_delegation_response = 11;
  optional ObtainObjectMessage obtain_object = 12;
  optional StorageDiscoverMessage storage_discover = 13;
  optional StorageHeartBeatMessage storage_heartbeat = 14;
  optional ObjectInfoMessage object_info = 15;
  optional TaskKillMessage task_kill = 16;
  optional TaskFinalReport task_final_report = 17;
}
